CAS-2081: use cas3 v2 endpoints

froddd opened the pull request:
# Context

https://dsdmoj.atlassian.net/browse/CAS-2081

# Changes in this PR

This PR enables the use of the new `/cas3/v2` API endpoints via the introduction of the `ENABLE_CAS3V2_API` feature flag.

The codebase was fully migrated to the newer types in order to be run wither with the flag off or on; this means that any entities of 'older' types coming from the non-v2 endpoints (and, in many other cases, even older endpoints) are now cast as the newer types, so everything will continue to work during the transition period.

The approach to this work has largely used the following process:

- Remove usages of older types and replace them with the newer types, as returned by the newer endpoints. This has introduced a lot of casting utilities, which will be removed in a subsequent PR when the migration has been fully deployed.
- Update clients to point to either the 'old' endpoints, or the newer `/cas3/v2` ones. This means the clients may return different types while the feature flag is actionable.
- Update services to cast older types as the newer ones. This includes any `POST` or `PUT` requests for creating/updating entities: the whole codebase now uses the newer request types, too.
- Update controllers and templates to use the newer types/entity structures.

Other than updating types and factories, the E2E and integration tests have been left largely untouched, to ensure the changes are not affecting functionality. Some exceptions to this:

- Reference data for premises and bedspace characteristics has been consolidated and is now based on the actual records available on preprod. This ensures the tests can use consistent data throughout the environments.
- a number of the mock APIs for integration tests now contain 2 versions of each endpoints: one for the `/cas3v2` endpoints, and another for the older ones.


I have tried to avoid making ad-hoc improvements to the implementation itself: some areas may seem like obvious candidates, but as this is a large refactor, it seemed a better idea to touch as little (!) of teh codebase as possible.

Where possible, utilities and conditionals thet enable the codebase to work with or without the feature flag enabled have been preceded with a `// TODO -- ENABLE_CAS3V2_API cleanup` comment -- this may help in determining what parts of the code should be removable once this work is deployed. However, I will also prepare a separate PR containing full removal of all conditionals.

## E2E tests

Both Cypress and PlayWright E2E tests are passing locally with the feature flag set to `false` or `true`.

## Screenshots of UI changes

This PR does not introduce any UI changes.

# Release checklist

[Release process documentation](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/edit-v2/4247847062?draftShareId=a1c360ab-bd31-4db1-aae3-7cc002761de9).

## Pre-merge checklist

- [ ] Are any changes required to dependent services for this change to work?
  (eg. CAS API)
- [ ] Have they been released to production already?

## Post-merge checklist

- [ ] [Manually approve](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/4247847062/Release+process#Manual-releases)
  release to test
- [ ] [Manually approve](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/4247847062/Release+process#Manual-releases)
  release to preprod
- [ ] [Manually approve](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/4247847062/Release+process#Manual-releases)
  release to prod

<!-- Should a release fail at any step, you as the author should now lead the work to
fix it as soon as possible. You can monitor deployment failures in CircleCI
itself and application errors are found in
[Sentry](https://ministryofjustice.sentry.io/issues/?project=4504129156218880&referrer=sidebar&statsPeriod=24h).
Both events should be automatically sent to our [Slack
channel](https://mojdt.slack.com/archives/C048BJS7S2F). -->

froddd committed:
Clean up API endpoint paths definitions

This removes any secondary variable assignments which were just making
the file very difficult to read. The only set of endpoints that still
has its own variable is the manage CAS3 paths, as these will be
overwritten where necessary by the newer, cas3v2 endpoints.

froddd committed:
Move booking endpoints to api definition file

froddd committed:
Fix query logging for GET requests

froddd committed:
Move premises utilities out of service

froddd committed:
Reorganise api definitions to introduce cas3v2

This renames a lot of the transitory variables to be prefixed with
`cas3` or `cas3v2` where necessary, mostly for readability. It also
moves any path that is already querying the `/cas3` namespace on the
api to the `cas3` property, again, for clarity.

Finally, this introduces the feature-switchable `/cas3/v2` prefix for
premises and bedspaces endpoints that are already available on this
namespace.

froddd committed:
Move bedspace utilities out of service

froddd committed:
Ensure E2E have access to a process.env

This is necessary as the API paths are now dependent on the main config,
which calls on process.env for various config flags. We simply pass an
empty object to ensure Cypress does not bail.

froddd committed:
Replace usages of Booking type with Cas3Booking

This commit introduces a type transformer used on responses from the
API: objects of type Booking are transformed into objects of type
Cas3Booking. The transformers also accept objects of type Cas3Booking as
input, in which case they just pass these through: this will allow us
to switch the API endpoints used from those that return Booking types to
those that return Cas3Booking types transparently.

API endpoints have been updated to enable the CAS3v2 version when the
feature flag is enabled.

froddd committed:
Replace usages of LostBed with Cas3VoidBedspace

Also replaces related types. The response from the API is cast as
Cas3VoidBedspaces so the codebase only uses the latest type, to
facilitate a later cleanup of deprecated types.

froddd committed:
Fix small issue with characteristics

froddd committed:
Run integration tests with feature flag off and on

froddd committed:
Replace usages of BedspaceSearchResults

These are replaced with the newer type Cas3v2BedspaceSearchResults, and
all other linked types. As per previous commits for Lostbed and Booking
types, the response from the API is now transformed to an object of
type Cas3v2BedspaceSearchResults.

froddd committed:
Move bookingSearch casting utilities

froddd committed:
Improve Cypress a11y errors logging

This ensures the actual a11y violations are logged
to the Cypress command log.

froddd committed:
Use new characteristics key and reference data

This replaces the use of the deprecated property `characteristics`
in both the Cas3Premises and Cas3Bedspace types.

The reference data used for generating sets of characteristics
radio buttons is now fetched from the CAS3-namespaced API endpoint
if the ENABLE_CAS3V2_API flag is enabled.

froddd committed:
Increase unrelated tests coverage

froddd committed:
Fix test reference data

The reference data is copied from what is currently available on
preprod. Only reference data that is marked as active has been added,
with the exception of the 'other' option, which is filtered in the UI
but shoudl probably be marked as inactive in the DB.

froddd committed:
Add last few missing CAS3v2 endpoints

froddd committed:
Correctly handle void bedspace conflicts

froddd commented on integration_tests/mockApis/cancellation.ts:
```diff
@@ -14,25 +16,52 @@ export default {
       request: {
-        method: 'GET',
-        url: `/premises/${args.premisesId}/bookings/${args.bookingId}/cancellations/${args.cancellation.id}`,
+        method: 'POST',
```
The diff view is a bit out of whack here -- this should map to the old `stubCancellationCreate` above!

froddd commented on integration_tests/tests/temporary-accommodation/manage/bedspace.cy.ts:
```diff
@@ -487,21 +492,11 @@ context('Bedspace', () => {
-        },
+      const bedspaceCharacteristics: Array<Cas3BedspaceCharacteristic> = [
+        cas3BedspaceCharacteristicsFactory.byDescription('Ground floor level access'),
+        cas3BedspaceCharacteristicsFactory.byDescription('Shared kitchen'),
```
The factory for characteristics now returns more realistic characteristics based on the ones available on preprod at the moment: 'Single bed' didn't actually exist for bedspaces, so it's been swapped with another that does exist.

froddd commented on integration_tests/tests/temporary-accommodation/manage/booking.cy.ts:
```diff
@@ -143,8 +142,12 @@ context('Booking', () => {
 
-      expect(requestBody.service).equal('temporary-accommodation')
-      expect(requestBody.bedId).equal(bedspace.id)
+      if (Cypress.env('ENABLE_CAS3V2_API') === true) {
```
Cypress doesn't like to access `config.flags` when running the integration tests, so instead we have to rely on this extra Cypress environment variable. The Github Actions workflow specifies this as `CYPRESS_ENABLE_CAS3V2_API`. The extra workflow will be fully removed when this is deployed.

froddd commented on server/controllers/temporary-accommodation/manage/bedspacesController.test.ts:
```diff
@@ -45,8 +47,13 @@ describe('BedspacesController', () => {
+
   const referenceData = {
-    characteristics: referenceDataFactory.characteristic('room').buildList(5),
+    characteristics: [...displayedCharacteristics, cas3ReferenceDataFactory.build({ description: 'Other' })],
```
There are some conditionals used when rendering the radio buttons that filter out the 'Other' option -- this wasn't explicitly tested, so this now covers it. I don't actually think the 'Other' option is shown or rendered anywhere, so it may be possible and/or desirable to set it to `active=false` in the DB?

froddd commented on server/controllers/temporary-accommodation/manage/bedspacesController.test.ts:
```diff
@@ -233,39 +240,13 @@ describe('BedspacesController', () => {
-      characteristics: [characteristic1, characteristic2],
+      bedspaceCharacteristics: [characteristic1, characteristic2],
     })
-    const onlineBedspaceSummary: SummaryList = {
```
This output is tested in its own utility test suite, so we can simply assert it gets called with the right arguments.

froddd commented on server/controllers/temporary-accommodation/manage/bookingsController.ts:
```diff
@@ -207,19 +208,16 @@ export default class BookingsController {
       const { arrivalDate } = DateFormats.dateAndTimeInputsToIsoString(req.body, 'arrivalDate')
       const { departureDate } = DateFormats.dateAndTimeInputsToIsoString(req.body, 'departureDate')
 
-      const bedspace = await this.bedspacesService.getSingleBedspace(callConfig, premisesId, bedspaceId)
```
The bedspace wasn't actually used for anything other than its `id`, which we already have as `bedspaceId` from the path params, so we save a call to the API here.

froddd commented on server/controllers/temporary-accommodation/manage/premisesController.test.ts:
```diff
@@ -75,42 +86,25 @@ describe('PremisesController', () => {
   })
 
   describe('index', () => {
-    const premisesRow1 = [
```
`tableRows` is covered in its own test suite, so no need for this fixture data here.

froddd commented on server/controllers/temporary-accommodation/manage/premisesController.test.ts:
```diff
@@ -120,7 +114,7 @@ describe('PremisesController', () => {
       })
 
-      premisesService.searchDataAndGenerateTableRows.mockResolvedValue(searchData)
+      premisesService.search.mockResolvedValue(searchData)
```
The clients have largely been stripped of any formatting/transforming duties, to make it easier to switch between testing with flag on and off -- these transforms have been moved either to the respective services, or to separate utilities.

froddd commented on server/data/bedspaceClient.test.ts:
```diff
@@ -196,4 +205,38 @@ describeClient('BedspaceClient', provider => {
     })
   })
+
+  describe('with the ENABLE_CAS3V2_API flag off', () => {
```
The tests contained in the `with the ENABLE_CAS3V2_API flag off` describe blocks are as close to what the original tests were as possible, to ensure no functionality changes. Unfortunately the diff will show them all as new additions... I have tried to move things around to produce a better diff, but no luck. Soz 🤷‍♂️

froddd commented on server/data/bedspaceClient.ts:
```diff
@@ -18,8 +19,15 @@ export default class BedspaceClient {
   async search(searchParameters: Cas3BedspaceSearchParameters) {
-    return this.restClient.post<Cas3BedspaceSearchResults>({
-      path: paths.bedspaces.search({}),
+    if (!config.flags.enableCas3v2Api) {
+      return this.restClient.post<Cas3BedspaceSearchResults>({
+        path: paths.cas3.bedspaces.search({}),
+        data: searchParameters,
+      })
+    }
```
Using a negative conditional here should make removal of the old branches a lot easier. The new world is the default now!

froddd commented on server/paths/api.ts:
Quite a bit of a refactor here: it was getting difficult to parse which paths were CAS3-namespaced and which ones were not, given a lot of the transitory objects setup for `assessPaths`, `applicationPaths`, etc. I decided to flatten things a fair bit and it did help during development -- with the unfortunate side effect that it makes this diff difficult to parse! I'm hoping the recent update to the client tests, bringing in proper contract testing, will remove any nervousness about the validity of all the paths defined here 🙂

SaintRhys commented on integration_tests/mockApis/extension.ts:
```diff
@@ -4,8 +4,63 @@ import type { Extension } from '@approved-premises/api'
+import config from '../../server/config'
 
-export default {
+const cas3V2ApiEnabledStubs = {
```
Minor but should this be `cas3v2ApiEnabledStubs`

froddd replied:
Eagle eyes! Thanks 🙂

SaintRhys commented on server/testutils/factories/cas3Booking.ts:
```diff
@@ -0,0 +1,135 @@
+  const turnaroundStartDate = faker.date.soon({ days: 1, refDate: departureDate })
+  const effectiveEndDate = faker.date.soon({ days: 5, refDate: departureDate })
+
+  const cancellations = faker.helpers.arrayElements(cancellationFactory.buildList(5))
```
Could these be using cas3 factories

froddd replied:
Yes, good spot! They absolutely should, in fact. I'm surprised the typecheck isn't actually picking this up 🤔

SaintRhys commented on e2e/tests/stepDefinitions/manage/booking.ts:
```diff
@@ -31,7 +31,7 @@ Given('I create a booking with all necessary details', () => {
       })
 
-      const booking = bookingFactory.provisional().build({
+      const booking = cas3BookingFactory.provisional().build({
```
Could the below be cas3 factory for turnaround

froddd replied:
Absolutely, fixed!

SaintRhys commented on server/testutils/factories/cas3Premises.ts:
```diff
@@ -65,7 +67,8 @@ export default Cas3PremisesFactory.define(() => ({
   totalArchivedBedspaces: faker.number.int({ min: 1, max: 5 }),
   totalUpcomingBedspaces: faker.number.int({ min: 1, max: 5 }),
-  characteristics: characteristicFactory.buildList(5),
+  characteristics: undefined as Array<Characteristic>,
```
Would this be better as an empty array or is it needed as undefined

froddd replied:
The new API doesn't return this property at all, so `undefined` is correct (it populates the `premisesCharacteristics` instead). But you make a good point -- it should probably not be set at all, so I'm removing it fully, and also removing the equivalend in the bedspace factory.

SaintRhys reviewed (commented):
Looking good, some solid changes.

There's a `Booking` in `bookingSearch` (cypress_shared) that could be swapped to `Cas3Booking`

and a couple bits about old factories in new ones.

froddd commented on server/utils/bedspaceUtils.test.ts:
```diff
@@ -142,4 +152,308 @@ describe('bedspaceV2Utils', () => {
     })
   })
+
+  describe('summaryList', () => {
```
The Summary List-related tests below have been moved from the test suite for the Bedspace Service, where the implementation originally resided.

froddd commented on server/utils/bedspaceUtils.ts:
```diff
@@ -133,3 +141,141 @@ export function setDefaultStartDate(userInput: Record<string, unknown>) {
   }
 }
+
+export const summaryList = (bedspace: Cas3Bedspace): SummaryList => {
```
These methods have been moved from the Bespace Service.

froddd commented on server/utils/premisesUtils.test.ts:
```diff
@@ -218,4 +234,408 @@ describe('premisesUtils', () => {
     })
   })
+
+  describe('tableRows', () => {
```
Tests and implementations moved from the Premises Service.

froddd committed:
Add feature flag to helm charts

froddd committed:
Use unique references for bedspace search results

In some rare casess, multiple search results as generated for the
integration tests may have had references starting with the same
numbers, as the references themselves were chosen as random numbers
between 1 and 9999.

For instance, one result could have had reference '81', with
another having reference '8190'. In this case, the integration test
searching for teh result with reference '81' may have returned the
result container for the one with reference '8190', breaking the
assertions further down in the test.

To fix this, we use faker's alphanumeric string generator for results
reference, to minimise the chance of collisions.

froddd committed:
Split booking client tests and fix factories

This ensures the correct API endpoints are used in tests, and updates
all client calls to the V2 endpoints by updating factories and other
related artefacts to match the expected OpenAPI spec.

froddd committed:
Split bedspace client tests

This ensures the correct API endpoints are used in tests.

froddd committed:
Split LostBed client tests

This ensures the correct endpoints are used for tests.
